code/datums/components/material_container.dm

```diff
@@ -136,69 +173,106 @@
 	last_inserted_id = insert_item_materials(I, multiplier, breakdown_flags)
 	return material_amount
 
-/datum/component/material_container/proc/insert_item_materials(obj/item/I, multiplier = 1, breakdown_flags = mat_container_flags)
+/**
+ * Inserts the relevant materials from an item into this material container.
+ *
+ * Arguments:
+ * - [source][/obj/item]: The source of the materials we are inserting.
+ * - multiplier: The multiplier for the materials being inserted.
+ * - breakdown_flags: The breakdown bitflags that will be used to retrieve the materials from the source
+ */
+/datum/component/material_container/proc/insert_item_materials(obj/item/source, multiplier = 1, breakdown_flags = mat_container_flags)
 	var/primary_mat
 	var/max_mat_value = 0
-	var/list/item_materials = I.get_material_composition(breakdown_flags)
-	for(var/MAT in materials)
+	var/list/item_materials = source.get_material_composition(breakdown_flags)
+	for(var/MAT in item_materials)
+		if(!can_hold_material(MAT))
+			continue
 		materials[MAT] += item_materials[MAT] * multiplier
 		total_amount += item_materials[MAT] * multiplier
 		if(item_materials[MAT] > max_mat_value)
 			max_mat_value = item_materials[MAT]
 			primary_mat = MAT
+
 	return primary_mat
 
+/**
+ * The default check for whether we can add materials to this material container.
+ *
+ * Arguments:
+ * - [mat][/atom/material]: The material we are checking for insertability.
+ */
+/datum/component/material_container/proc/can_hold_material(datum/material/mat)
+	if(mat in allowed_materials)
+		return TRUE
+	if(istype(mat) && ((mat.id in allowed_materials) || (mat.type in allowed_materials)))
+		allowed_materials += mat	// This could get messy with passing lists by ref... but if you're doing that the list expansion is probably being taken care of elsewhere anyway...
+		return TRUE
+	if(insertion_check?.Invoke(mat))
+		allowed_materials += mat
+		return TRUE
+	return FALSE
```

Comment:
Early returns would be good here

Reply:
I don't see what you mean, I'm already returning early.

Reply:
brain empty no thoughts